capstone/fitness/src/main/java/com/google/sps/util/BananaQueue.java

```diff
@@ -0,0 +1,204 @@
+    if(head == null || head.peelQueueExists(peelQueueTag)) {
+      return false;
+    }
+
+    head.addPeelQueue(peelQueueTag, peelQueue);
+    return true;
+  }
+
+  /**
+   * Removes PeelQueue from the front element of the BananaQueue with the given tag. 
+   *
+   * @param peelQueueTag Tag to access PeelQueue.
+   * @return removed PeelQueue
+   */
+  public PeelQueue removePeelQueue(String peelQueueTag) {
+    if(head == null || !head.peelQueueExists(peelQueueTag)) {
+      return null;
+    }
+
+    return head.removePeelQueue(peelQueueTag);
+  }
+
+  /**
+   * Adds BananaNode to the end of the queue and increments the size.
+   * 
+   * @param banana BananaNode to add
+   * @return true if enqueue is successful
+   */
+  public boolean enqueueBanana(BananaNode banana) {
+    if(banana == null) {
+      return false;
+    }
+
+    if(head == null) {
+      head = banana;
+    }
+    else { 
+      BananaNode temp = foot != null ? foot : head;
+      foot = banana;
+      temp.setNext(foot);
+      foot.setPrev(temp);
+    }
+
+    size++;
+    return true;
+  }
+  
+  /**
+   * Marks the front BananaNode in the queue as complete, decrements the size, moves the 
+   * front of the queue to the next element (can be null), and copies all the PeelQueues 
+   * in the old front to the new front (if not null).
+   *
+   * @return dequeued BananaNode
+   */
+  public BananaNode dequeueBanana() {
+    if(head == null) {
+      return null;
+    }
+
+    head.setComplete(true);
+    if(head.getNext() != null) {
+      head.getNext().addPeels(head.getPeels());
+    }
+    head.removePeels();
+    BananaNode temp = head;
+    head = head.getNext();
+    size--;
+    return temp;
+  }
+
+  /**
+   * Returns the front BananaNode of the queue (can be null).
+   * 
+   * @return front of BananaQueue.
+   */
+  public BananaNode peekBanana() {
+    return head;
+  }
+
+  /**
+   * Returns the front PeelNode of the queue (can be null) given an associated tag.
+   *
+   * @param peelQueueTag Tag to associated PeelQueue.
+   * @return front of PeelQueue mapped to tag
+   */
+  public PeelNode peekPeel(String peelQueueTag) {
+    if(head == null || !head.peelQueueExists(peelQueueTag)) {
+      return null;
+    }
+    return head.getPeelQueue(peelQueueTag).peek(); 
+  }
+  
+  /**
+   * Returns the number of BananaNodes in the BananaQueue.
+   * 
+   * @return number of BananaNodes in the BananaQueue.
+   */
+  public int getSize() {
+    return size;
+  }
+
+  /**
+   * Returns the number of PeelNodes in the PeelQueue of the associated tag. If not found, 
+   * then it returns "-1".
+   *
+   * @param peelQueueTag Tag to associated PeelQueue.
+   * @return number of PeelNodes in the PeelQueue mapped to tag.
+   */
+  public int getPeelSize(String peelQueueTag) {
+    if(head == null || !head.peelQueueExists(peelQueueTag)) {
+      return -1;
+    }
+    return head.getPeelQueue(peelQueueTag).getSize();
+  }
+  
+  public void setSize(int size) {
```

Comment:
Do you need the setSize function?

Reply:
I was thinking of needing it in the context of the progress model instead of the actual BananaQueue because I thought if a user wanted to revert the model, there'd be no way. However, this probably isn't needed/ruins the idea of the data structure. An alternative can be creating a whole new model based off the last one if a user wants to revert.

Reply:
Yeah I think this type of function doesn't make as much sense for the BananaQueue data structure so if we decide we want to have a revert we can revisit how to do that later.